airbyte-integrations/connectors/source-mongodb/lib/mongodb_source.rb

```diff
@@ -0,0 +1,81 @@
+require_relative './airbyte_protocol.rb'
+require_relative './airbyte_logger.rb'
+
+require_relative './mongodb_stream.rb'
+require_relative './mongodb_reader.rb'
+
+class MongodbSource
+  def spec
+    spec = JSON.parse(File.read(__dir__ + '/spec.json'))
+
+    message =  AirbyteMessage.from_dynamic!({
+      'type' => Type::Spec,
+      'spec' => spec,
+    })
+
+    puts message.to_json
+  end
+
+  def check(config:)
+    @config = JSON.parse(File.read(config))
+
+    result = {'status' => Status::Failed, 'message' => 'Something went wrong.'}
+
+    begin
+      client.collections.first.find.limit(1).first
+      result = {'status' => Status::Succeeded}
+    rescue Mongo::Auth::Unauthorized => e
+      AirbyteLogger.log(e.backtrace.join("\n"), Level::Fatal)
+      result = {'status' => Status::Failed, 'message' => 'Authentication failed.'}
+    end
+
+    message =  AirbyteMessage.from_dynamic!({
+      'type' => Type::ConnectionStatus,
+      'connectionStatus' => result,
+    })
+
+    puts message.to_json
+  end
+
+  def discover(config:)
+    @config = JSON.parse(File.read(config))
+
+    streams = client.collections.map do |collection|
+      MongodbStream.new(collection: collection).discover
+    end
+
+    catalog = AirbyteCatalog.from_dynamic!({
+      'streams' => streams,
+    })
+
+    puts AirbyteMessage.from_dynamic!({
+      'type' => Type::Catalog,
+      'catalog' => catalog.to_dynamic
+    }).to_json
+  end
+
+  def read(config:, catalog:, state: nil)
+    @config = JSON.parse(File.read(config))
+    @catalog = JSON.parse(File.read(catalog))
+    @state = JSON.parse(File.read(state)) if state
+
+    MongodbReader.new(client: client, catalog: @catalog, state: @state).read
+  end
+
+  def method_missing(m, *args, &block)
+    AirbyteLogger.log("There's no method called #{m}", Level::Fatal)
+  end
+
+  private
+
+  def client
+    if @client
```

Comment:
might be simpler to do `if not @client` and run the else block, then always return `@client`

Reply:
I've reworked that memoization pattern in a bit different way. Please let me know if that approach will work